### PR TITLE
Re-port `selectionsToBabel` -> `selectionsToAST` to fixup union type decls

### LINKED ISSUE #120: Context TypeName not a union
I'm seeing a bug with `Context.__typeName` evaluating as a string instead of a union, seen here: https://github.com/artsy/emission/pull/1685/files#diff-7b710d59e493e84a14f4a395ffb8d3ceR8; resulting in typer errors necessitating [this line](https://github.com/artsy/emission/pull/1685/files#diff-aceb5ae87ee50364fe8f6a7b4892a1afR52) to be typed as `any`.

### PR DESCRIPTION
**Pull request by kastermester**
See the changes in the tests. This should allow much easier work with
types that can take many different shapes.

I assume this will fix #120.

### CODE REVIEW

**Comment by kastermester** [comment]
I will need to do some more testing with our codebase to make sure this actually has the intended behavior, will attempt to report back on monday.

**Comment by kastermester** [comment]
After some further investigation I decided to port all the functions again.

I've run some tests and this seems to have fixed some issues for us - as well as matching the behavior of the flow generation now.

Please note if you do the following:

```graphql
Fragment SomeFragment on SomeUnionType {
  __typename
  ... on SomeConcreteType {
    someField: {
     ... SomeFieldFragment
   }
  }
  ... on SomeOtherConcreteType {
    someField: {
     ... SomeFieldFragment
   }
  }
  ... SomeSecondFragment
  ... SomeThirdFragment
}
```

Then the generated types will look something like this:

```ts
    readonly __typename: string;
    readonly someField?: {
        readonly " $fragmentRefs": SomeFieldFragment$ref;
    };
    readonly " $fragmentRefs": SomeSecondFragment$ref & SomeThirdFragment$ref;
    readonly " $refType": SomeFragment$ref;
```

The solution to get proper types is to move the fragment spreads that are not under a type guard into the type guards:

```graphql
Fragment SomeFragment on SomeUnionType {
  __typename
  ... on SomeConcreteType {
    someField: {
     ... SomeFieldFragment
   }
    ... SomeSecondFragment
    ... SomeThirdFragment
  }
  ... on SomeOtherConcreteType {
    someField: {
     ... SomeFieldFragment
   }
    ... SomeSecondFragment
    ... SomeThirdFragment
  }
}
```

Which should generate these types:

```ts
{
    readonly __typename: "SomeConcreteType";
    readonly someField: {
        readonly " $fragmentRefs": SomeFieldFragment$ref;
    };
    readonly " $fragmentRefs": SomeSecondFragment$ref & SomeThirdFragment$ref;
    readonly " $refType": SomeFragment$ref;
} | {
    readonly __typename: "SomeOtherConcreteType";
    readonly someField: {
        readonly " $fragmentRefs": SomeFieldFragment$ref;
    };
    readonly " $fragmentRefs": SomeSecondFragment$ref & SomeThirdFragment$ref;
    readonly " $refType": SomeFragment$ref;
} | {
    /*This will never be '%other', but we need some
    value in case none of the concrete values match.*/
    readonly __typename: "%other";
    readonly " $refType": SomeFragment$ref;
}
```

I would consider this less than ideal - however according to my tests, the flow type generation behaves the same. I would very much appreciate some extra eyes on the test snapshots. I've checked some of them as well as ran this on our own (note: this is still a tiny codebase wrt. relay modenr for now) codebase and it seems to work as intended.

File: test/TypeScriptGenerator-test.ts [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
 
 function generate(text, options) {
-  const schema = transformASTSchema(RelayTestSchema, [
-    IRTransforms.schemaExtensions[1], // RelayMatchTransform.SCHEMA_EXTENSION,
-    IRTransforms.schemaExtensions[2], // RelayRelayDirectiveTransform.SCHEMA_EXTENSION,
-    `
+	const schema = transformASTSchema(RelayTestSchema, [
+		IRTransforms.schemaExtensions[1], // RelayMatchTransform.SCHEMA_EXTENSION,
+		IRTransforms.schemaExtensions[2], // RelayRelayDirectiveTransform.SCHEMA_EXTENSION,
+		`
```
Comment (alloy) [inline test/TypeScriptGenerator-test.ts]: Whitespace gone wild? 😄 

Perhaps run `yarn prettier`? I'm surprised the git hook didn’t automatically do this, though 🤔

Comment (kastermester) [reply test/TypeScriptGenerator-test.ts]: I'm quite certain I did exactly this, but let me try again.

Comment (alloy) [reply test/TypeScriptGenerator-test.ts]: Ah, looks like `./test` isn’t included in the glob https://github.com/relay-tools/relay-compiler-language-typescript/blob/master/package.json#L38

**Comment by alloy** [comment]
At a glance this is looking like a step in the right direction 👌 But just judging by the snapshots is hard, so I'll have to run this against our codebase to see how it pans out. Seeing as this is my first day back from vacation and I have a bunch of things to catch-up on, it may take a few days.

File: test/TypeScriptGenerator-test.ts (hunk 1 above)
Comment (kastermester) [reply test/TypeScriptGenerator-test.ts]: Right - I have changed this and included it in the PR as well.

**Comment by alloy** [comment]
Just had the time to try this on our codebase and it indeed fixes #120, as well as generally improves union types. Great stuff!

**Comment by alloy** [comment]
Going to go with a major bump, as I expect upgrading to this version will lead to some broken builds where tsc finds new issues.
